### DISCUSSION TITLE
Make type family applications in instances legal if they can be fully evaluated

### DISCUSSION THREAD

**Discussion by JakobBruenker** [ghc-proposals/ghc-proposals#463] [Ideas]
I just had this idea, so it's possible that I'm missing something, or that this has already been discussed elsewhere. Hence I'm opening a discussion to get some feedback.

### Motivation

I just tried to write this instance:
```haskell
instance Escapable (LHsExpr GhcPs) where
```
Since `LHsExpr` (from the GHC API) is a type family, GHC rightly complains that this is an "Illegal type synonym family application".

It can be worked around easily: We can ask GHCi what the application evaluates to
```haskell
ghci> :kind! LHsExpr GhcPs         
LHsExpr GhcPs :: *
= GenLocated
    (SrcSpanAnn' (EpAnn AnnListItem)) (HsExpr (GhcPass 'Parsed))
```
since there is no stuck type family application left, we can use this to get
```haskell
instance Escapable (GenLocated (SrcSpanAnn' (EpAnn AnnListItem)) (HsExpr (GhcPass 'Parsed))) where
```
This works - but the resulting instance head is unwieldy and hard to understand.

### Proposed Changes

As long as -XTypeSynonymInstances is enabled, GHC should allow instances that contain only type family applications that can be fully evaluated. It should behave the same way as if the instance were defined with the fully evaluated form.

At the moment, I can't see anything that would make this particularly difficult.

**Comment by nomeata** [comment]
There is an underlying question here, namely: where do we treat type synonyms differently from fully applied (and fully evaluatable) type family applications different, and is that desirable. (In my mental model, the two are quite similar, so some cleanup here might be appreciated.)

**Comment by adamgundry** [comment]
I think this would be perfectly feasible to implement. The question is whether we want it. After all, we don't allow it in patterns at the value level, although something like this would be equally meaningful:
```hs
f (not True)  = True
f (not False) = False
```

This came up in a discussion around #458, because if we had this feature, we could have `Tuple [Int, Bool, Double]` be a type family that reduced to `Tuple3 Int Bool Double` and yet still be able to write an instance like `instance C (Tuple [Int, Bool, Double])`.

**Reply by JakobBruenker** [reply]
Hm, that's a good point - I hadn't considered that you'd likely want it to work with type family instances, too (rather than just class instances).

I suppose while functions are the term-level analogue to type families, pattern synonyms work somewhat analogously to type synonyms.

**Reply by JakobBruenker** [reply]
In my mind this is a decent argument not to pursue this idea, but perhaps there is one related thing we could do: Allow type family applications that can be fully evaluated to appear in the right-hand side of an injective type family (I stumbled over this when adjusting the code I used as motivating example here).

**Comment by goldfirere** [comment]
I've sometimes wondered about

```hs 
instance C a | a ~ F Int where ...   -- version (1)
```

This is distinct from

```hs
instance a ~ F Int => C a where ...   -- version (2)
```

My new (1) checks the constraint *before* committing to the instance. On the other hand, the old (2) commits to the instance first and then tries to satisfy the constraint.

In order to keep type checking from backtracking, we would have to restrict the constraints after the `|` to include only those that can be checked quickly, without further solving. I think that, today, that includes only equality and, tantalizingly, apartness. That is, imagine

```hs
instance Different a b | a /~ b
```

This would be an alternative to using closed type families to control instance selection. Indeed, if we extend this treatment to type family instances (we should!), we can get rid of closed type families altogether. (I actually don't want this, because closed type families work just like functions. But if we ever had Constrained Type Families like #177, then this notation would work well for partial type families that need apartness guards.

Giving credit where it's due: this has been kicking around in my head since #56.
